simpledotorg/simple-server
# Implement authentication for stakeholder resource access

**Story card:** [sc-10740](https://app.shortcut.com/simpledotorg/story/10740)

## Context

[Link to relevant section of the PRD](https://github.com/doorkeeper-gem/doorkeeper/wiki)

Currently, we need to create OAuth Applications for users that want to send us data through the `/import` API.

Since our partner organizations are trusted machine clients in this case, we create and send them a client ID and secret that is used for the [client credential flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/client-credentials-flow) needed for the import API. We also create a machine user associated with these organizations.

We introduce a new dependency, [Doorkeeper](https://github.com/doorkeeper-gem/doorkeeper/wiki) to implement our OAuth provider.

The expiry token is set to 8 hours by default.

## This addresses

* Adds an OAuth provider, which involves migrations to hold OAuth-related details, all mostly generated by doorkeeper.
* Adds a machine user model, which currently only holds a name and organization.
* Secure the `/import` endpoint to only those who provide an OAuth token with the `write` scope.
* Adds a rake task to register and set up an OAuth application, whose details we will hand over to the partner orgs so they can integrate with us.

## Test instructions

Run the rake task (refer to the howto doc added in the PR).

Run the following HTTP request to obtain a token:
```
POST http://localhost:3000/oauth/token
Authorization: Basic <base64(client_id:client_secret)>
Content-Type: application/x-www-form-urlencoded;charset=UTF-8

grant_type=client_credentials
```

Copy the token ID, and call the API with it:
```
PUT http://localhost:3000/api/v4/import
Content-Type: application/json; charset=UTF-8
X-Organization-ID: <use the org ID passed into the rake task>
Authorization: Bearer <token>

{"resources": []} # use a valid resource if you want a 202 response instead of a 400
```

Any status code that's not 401 or 403 means the authorization worked. Conversely, an invalid token should return a 403 and no token at all should return a 401.
